internal/controller/etcdopstask/reconcile_task.go

```diff
@@ -0,0 +1,222 @@
+		return ctrlutils.ReconcileWithError(err)
+	}
+
+	if controllerutil.ContainsFinalizer(task, druidapicommon.EtcdOpsTaskFinalizerName) {
+		return ctrlutils.ContinueReconcile()
+	}
+
+	if err := kubernetes.AddFinalizers(ctx, r.client, task, druidapicommon.EtcdOpsTaskFinalizerName); err != nil {
+		if updateErr := r.updateTaskStatus(ctx, task, taskStatusUpdate{
+			Operation: &druidv1alpha1.LastOperation{
+				Type:        druidv1alpha1.LastOperationTypeReconcile,
+				State:       druidv1alpha1.LastOperationStateRequeue,
+				Description: fmt.Sprintf("ensureTaskFinalizer failed to add finalizer: %v", err),
+			},
+		}); updateErr != nil {
+			return ctrlutils.ReconcileWithError(errors.Wrapf(err, "failed to update status after error: %v", updateErr))
+		}
+		return ctrlutils.ReconcileWithError(err)
+	}
+	return ctrlutils.ContinueReconcile()
+}
+
+// transitionToPendingState sets the task.status.state to Pending if not already set.
+func (r *Reconciler) transitionToPendingState(ctx context.Context, logger logr.Logger, taskObjKey client.ObjectKey, _ handler.Handler) ctrlutils.ReconcileStepResult {
+	logger = logger.WithValues("step", "transitionToPendingState")
+	logger.Info("Executing step: transitionToPendingState")
+
+	task, err := r.getTask(ctx, taskObjKey)
+	if err != nil {
+		return ctrlutils.ReconcileWithError(err)
+	}
+
+	if task.Status.State != nil {
+		return ctrlutils.ContinueReconcile()
+	}
+	// Set the task state to Pending
+	// and update the LastTransitionTime.
+	pendingState := druidv1alpha1.TaskStatePending
+	if err := r.updateTaskStatus(ctx, task, taskStatusUpdate{
+		State: &pendingState,
+	}); err != nil {
+		return ctrlutils.ReconcileWithError(err)
+	}
+	return ctrlutils.ContinueReconcile()
+}
+
+// admitTask checks if the task is in a pending state.
+// If so, it updates the phase to admit and invokes the task handler's Admit method.
+// If the admit operation is in progress, it requeues the task.
+// If the admit operation fails, it updates the task state to Rejected.
+// If the admit operation succeeds, it updates the task.status.state to InProgress.
+// If the task is not in its Pending state, it skips the step.
+func (r *Reconciler) admitTask(ctx context.Context, logger logr.Logger, taskObjKey client.ObjectKey, taskHandler handler.Handler) ctrlutils.ReconcileStepResult {
+	logger = logger.WithValues("step", "admitTask")
+	logger.Info("Executing step: admitTask")
+
+	task, err := r.getTask(ctx, taskObjKey)
+	if err != nil {
+		return ctrlutils.ReconcileWithError(err)
+	}
+	// If the task is not in Pending state, skip the admit step.
+	if task.Status.State != nil && *task.Status.State != druidv1alpha1.TaskStatePending {
+		return ctrlutils.ContinueReconcile()
+	}
+
+	var etcdOpsTaskList druidv1alpha1.EtcdOpsTaskList
+	if err = r.client.List(ctx, &etcdOpsTaskList, client.InNamespace(task.Namespace)); err != nil {
+		return ctrlutils.ReconcileWithError(err)
+	}
+
+	for _, existingTask := range etcdOpsTaskList.Items {
```

Comment:
I had earlier in a call recommended to create a function whose purpose is to find if there is an existing not-completed task for the same etcd and for the same operation running. However this was not done. Now i am adding an explicit comment so that it is not missed this time.